layers/matmul.py

```diff
@@ -30,4 +32,158 @@ def LlamaMLP(hidden_states, w_up, w_gate, w_down):
     return down_proj
 
 def lm_head(x, weight, bias=None):
-    return MLP(x, weight, bias)
\ No newline at end of file
+    return MLP(x, weight, bias)
+
+    
+class LlamaMLP_origin(nn.Module):
+    # from transformers modeling_llama.py
+    def __init__(self, hidden_size, intermediate_size, mlp_bias=None):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.intermediate_size = intermediate_size
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=mlp_bias)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=mlp_bias)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=mlp_bias)
+        self.act_fn = nn.functional.silu
+
+    def forward(self, x):
+        gate_proj = self.gate_proj(x)
+        up_proj = self.up_proj(x)
+        activated_gate_proj = self.act_fn(gate_proj)
+        intermediate_states = activated_gate_proj * up_proj
+        down_proj = self.down_proj(intermediate_states)
+        return down_proj
+
+
+def MLP_backward(grad_output, x, w_up, w_gate, w_down):
```

Comment:
- rename to LlamaMLP_backward, MLP means matmul  with/without bias
- add a new func: matmul_backward